packages/next/src/build/webpack/loaders/next-flight-loader/action-validate.ts

```diff
@@ -6,7 +6,22 @@ export function ensureServerEntryExports(actions: any[]) {
     const action = actions[i]
     if (typeof action !== 'function') {
       throw new Error(
-        `A "use server" file can only export async functions, found ${typeof action}.`
+        `A "use server" file can only export async functions, found ${typeof action}.\nRead more: https://nextjs.org/docs/messages/invalid-use-server-value`
+      )
+    }
+
+    if (action.constructor.name !== 'AsyncFunction') {
+      const actionName: string = action.name || ''
+
+      // Note: if it's from library code with `async` being transpiled to returning a promise,
```

Comment:
This bit is actually quite annoying for library authors tho, any plan where this will get improved in the future?

Reply:
Agree. It also breaks the use of Zod for validating the input of Server Actions (this worked before):

```ts
"use server";

import { db } from "./db";
import { z } from "zod";

export const getUserName = z
  .function()
  .args(z.object({ userId: z.string() }))
  .implement(async ({ userId }) => {
    const user = await db.users.query({ where: { id: userId } });
    return user.name;
  });
```